rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,213 @@
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `T::IteratorWrapper::into_pointer`.
+    // We already checked for he null pointer case above.
+    let mut iterator = unsafe { T::IteratorWrapper::from_pointer(v) };
+
+    // SAFETY: The caller guarantees tha `pos` is a valid pointer to an
+    // `loff_t` and expects this function to mutate the value.
+    unsafe {
+        *pos += 1;
+    }
+
+    if iterator.next().is_none() {
+        return ptr::null_mut();
+    }
+
+    match iterator.peek() {
+        Some(_next) => T::IteratorWrapper::into_pointer(iterator) as *mut _,
+        None => ptr::null_mut(),
+    }
+}
+
+extern "C" fn show_callback<T: SeqOperations>(
+    m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+) -> c_types::c_int {
+    const FORMAT: CStr<'static> = cstr!("%.*s");
+    if v.is_null() {
+        return 0;
+    }
+    // SAFETY: `v` was created by a previous call to `next_callback` or
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `T::IteratorWrapper::into_pointer`. We
+    // checked for null pointers above. The iterator is forgotten below
+    // so the pointer on the C side stays valid.
+    let mut iterator = unsafe { T::IteratorWrapper::from_pointer(v) };
+    if let Some(item) = iterator.peek() {
+        let s = T::display(item);
+        // SAFETY: Calling a C function. `FORMAT` is null terminated because
+        // it comes from a `CStr`. `s` does not need to be null terminated
+        // because we are only printing the first `s.len()` bytes.
+        unsafe {
+            bindings::seq_printf(
+                m,
+                FORMAT.as_ptr() as *const c_types::c_char,
+                s.len(),
+                s.as_ptr() as *const u8 as *const c_types::c_char,
+            );
+        }
+    }
+    // Need to forget the iterator because the C side is still using a
+    // reference to it.
+    mem::forget(iterator);
+    0
+}
+
+extern "C" fn start_callback<T: SeqOperations>(
+    m: *mut bindings::seq_file,
+    pos: *mut bindings::loff_t,
+) -> *mut c_types::c_void {
+    // SAFETY: This function will be called by opening a proc file generated
+    // from `proc_create_seq_private` on the C side with data created via
+    // `T::DataWrapper::into_pointer`. We don't move the data in the wrapper
+    // so the pointer will remain valid for later calls.
+    let data_wrapper =
+        unsafe { T::DataWrapper::from_pointer(bindings::PDE_DATA((*(*m).file).f_inode)) };
+    let iterator = T::start(&data_wrapper);
+    // Data is still used in the `proc_dir_entry`.
+    mem::forget(data_wrapper);
+    // SAFETY: The caller guarantees that `pos` points to a valid `loff_t`.
+    let pos = unsafe { *pos };
+    match iterator {
+        Some(mut wrapper) => {
+            for _ in 0..pos {
+                wrapper.next();
```

Comment:
I think here we need to do a similar check we're doing in `next_callback` to prevent potentially wrapping around and restarting iteration. That is, return null if `next` fails.

Reply:
yep, done.